crates/red_knot_python_semantic/src/semantic_index/visibility_constraints.rs

```diff
@@ -553,37 +554,99 @@ impl VisibilityConstraints {
         }
     }
 
+    fn analyze_single_pattern_predicate_kind<'db>(
+        db: &'db dyn Db,
+        predicate_kind: &PatternPredicateKind<'db>,
+        subject: Expression<'db>,
+    ) -> Truthiness {
+        match predicate_kind {
+            PatternPredicateKind::Value(value) => {
+                let subject_ty = infer_expression_type(db, subject);
+                let value_ty = infer_expression_type(db, *value);
+
+                if subject_ty.is_single_valued(db) {
+                    Truthiness::from(subject_ty.is_equivalent_to(db, value_ty))
+                } else {
+                    Truthiness::Ambiguous
+                }
```

Comment:
I think we can do disjointness checking for `Value` predicates too, right?

```suggestion
                } else if subject_ty.is_disjoint_from(db, value_ty) {
                    Truthiness::AlwaysFalse
                } else {
                    Truthiness::Ambiguous
                }
```

Reply:
Is this sound? With the suggested changes we have

```python
from typing import final

@final
class C:
    def __eq__(self, other: object) -> bool:
        return other == 1

def _(subject: C):
    y = 1
    match subject:
        case 1:
            y = 2
    reveal_type(y)  # revealed_type: Literal[1]
```

but I think (this is the same code but with `print` instead of `reveal_type`)

```python
from typing import final

@final
class C:
    def __eq__(self, other: object) -> bool:
        return other == 1

def _(subject: C):
    y = 1
    match subject:
        case 1:
            y = 2
    print(y)
```

```console
> uvx python /tmp/foo.py
2
```

Reply:
Ah, of course, great call. Yeah, we can't do this. It only works for single-valued subject (where we can be sure it doesn't compare equal to the predicate value.) And that's already supported.

Thanks for catching that!

Reply:
Seconded, thanks -- that's a great catch! I forgot how value patterns worked. Great that you added a test for this!